agent/src/beerocks/slave/tasks/task.h

```diff
@@ -0,0 +1,70 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#ifndef _TASK_H_
+#define _TASK_H_
+
+#include <beerocks/tlvf/beerocks_header.h>
+#include <tlvf/CmduMessageRx.h>
+
+namespace beerocks {
+
+class Task {
+public:
+    Task() {}
+    virtual ~Task() {}
+
+    /**
+     * @brief Enum which defines all tasks type.
+     * 
+     * Each task must have an entry on the enum eType, so it will be possible to send the task an
+     * event. 
+     */
+    enum eType : uint8_t {
+        TOPOLOGY,
+    };
+
+    /**
+     * @brief Retrun the task type.
+     * 
+     * Called by TaskPool send_event().
+     * 
+     * @return eType Task type.
+     */
+    virtual eType get_task_type() = 0;
+
+    /**
+     * @brief Work function. Shall do the task routine. Called by TaskPool run_task() function.
+     */
+    virtual void work(){};
+
+    /**
+     * @brief Handle events, which are being send from containing thread.
+     * 
+     * @param event_enum_value Event enum value which shall be defined on the task.
+     */
+    virtual void handle_event(uint8_t event_enum_value){};
+
+    /**
+     * @brief Handle CMDU message.
+     * 
+     * @param cmdu_rx CMDU object conataining the message.
+     * @param src_mac AL Mac of the sender.
+     * @param beerocks_header Beerocks Message header (Only on VS message).
+     * @return true if the message has been handled, otherwise false.
+     */
+    virtual bool handle_cmdu(ieee1905_1::CmduMessageRx &cmdu_rx, const sMacAddr &src_mac,
```

Comment:
Also, if by default, we fail to handle cmdu, why should we catch it only in run-time? 
Let's force the implementer to add a code here. In this way we prevent the problem instead of solving it later.
```suggestion
    virtual bool handle_cmdu(CmduHandleContext cmdu_context) = 0;
```

Reply:
The boolean returned value is for indicating that the message has been handled or not. Not for indicating that the handler function has failed or not.
That why it returns false by default, so if the Task derived class will not implement it, the task pool will continue iterating on other tasks, on TaskPool:handle_cmdu().

I have stated it on the Doxygen documentation.

Reply:
So how do you handle errors? Or the `TaskPool` doesn't care about errors?

Reply:
What is to handle? We print an error and stop the handling on the task. We have nothing to about it on the outside.
Also, if you follow the return value of the handlers functions across the code you will see that except for printing error we do nothing everywhere.

Reply:
> The boolean returned value is for indicating that the message has been handled or not. Not for indicating that the handler function has failed or not.

This is counter intuitive, exactly as it is today, and also as @vitalybu said, does not leave space for error-handling.
It also prevents more than one client to handle a message. When needed, patches are starting to pop like: "ok we'll return false here because we want the other task to be called as well".
The `return true means was handled` idiom should be replaced with something else, we failed with it.
One Option:
```suggestion
    // ret.first - handled/not handled
    // ret.second - success/fail
    // (true,true) - handled succefully. 
    // (true,false) - handled but error occurred 
    // (false,don't care) - not handled.
    virtual std::pair<bool,bool> handle_cmdu(ieee1905_1::CmduMessageRx &cmdu_rx, const sMacAddr &src_mac,
```

Another option, as @vitalybu suggested, some kind of registration to the CMDU that are with interest by the task.
In this way, the return value may be a simple `bool` indicating success/fail.